ffi/src/lib.rs

```diff
@@ -159,18 +275,35 @@ pub unsafe extern "C" fn store_transact(store: *mut Store, transaction: *const c
     Box::into_raw(Box::new(result.into()))
 }
 
+/// Fetches the `tx_id` for the given `TxReport`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TxReport` is not dangling.
 #[no_mangle]
 pub unsafe extern "C" fn tx_report_get_entid(tx_report: *mut TxReport) -> c_longlong {
     let tx_report = &*tx_report;
     tx_report.tx_id as c_longlong
 }
 
+/// Fetches the `tx_instant` for the given `TxReport`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TxReport` is not dangling.
 #[no_mangle]
 pub unsafe extern "C" fn tx_report_get_tx_instant(tx_report: *mut TxReport) -> c_longlong {
     let tx_report = &*tx_report;
     tx_report.tx_instant.timestamp() as c_longlong
 }
 
+/// Fetches the `Entid` assigned to the `tempid` during the transaction represented
+/// by the given `TxReport`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TxReport` is not dangling and that
+/// the C string provided to `tempid` is valid.
```

Comment:
What does "valid" mean here?  I see that we are manually handling `None => null` -- I rather hoped that `Option` would do that for us more directly.

And we're allocating memory to turn the `char *` into a Rust owned `String`, which we immediately borrow to look up a value in `tempids`.  What does it take to borrow the `char *` for the `tempids` lookup directly (crashing if we're given garbage)?  I'm pretty sure the `BTreeMap` under the hood here is designed for this type of flexibility when looking at the keys...

Reply:
Yup, you're right. We are doing unecessary work here. I'll see how we can improve this - string handling has been on my list to look at for a little while now.